### ISSUE TITLE
JSON::ParserError results in a failed build

### ISSUE THREAD

**Issue by rdh** [buildkite/test-collector-ruby#144]
Breaking this issue out from https://github.com/buildkite/test-collector-ruby/issues/138 where two different exception types were being discussed.

Over there, we added some extra logging on a branch to try and diagnose this one... https://github.com/buildkite/test-collector-ruby/issues/138#issuecomment-1191969426

Good news: The issue repro'd at 3:55am PT this morning.

Bad news:  The output is missing the log line reporting the JSON payload, which is what we were hoping to see.  It seems like the exception happened twice, but grepping the full GH Action output for "JSON", there's no sign of a "JSON Data" line.

```
2022-07-26T10:45:33.249275260Z pid=7771 tid=3354740 ERROR: JSON::ParserError: JSON::ParserError
#<Thread:0x0000560bb0507a00 /home/runner/work/masterclass/masterclass/backend/vendor/bundle/ruby/2.7.0/bundler/gems/test-collector-ruby-8074177c0faf/lib/buildkite/test_collector/socket_connection.rb:64 run> terminated with exception (report_on_exception is true):
/home/runner/work/masterclass/masterclass/backend/vendor/bundle/ruby/2.7.0/gems/rspec-mocks-3.10.2/lib/rspec/mocks/message_expectation.rb:143:in `block in and_raise': JSON::ParserError (JSON::ParserError)
	from /home/runner/work/masterclass/masterclass/backend/vendor/bundle/ruby/2.7.0/gems/rspec-mocks-3.10.2/lib/rspec/mocks/message_expectation.rb:694:in `block in call'
	from /home/runner/work/masterclass/masterclass/backend/vendor/bundle/ruby/2.7.0/gems/rspec-mocks-3.10.2/lib/rspec/mocks/message_expectation.rb:693:in `map'
	from /home/runner/work/masterclass/masterclass/backend/vendor/bundle/ruby/2.7.0/gems/rspec-mocks-3.10.2/lib/rspec/mocks/message_expectation.rb:693:in `call'
	from /home/runner/work/masterclass/masterclass/backend/vendor/bundle/ruby/2.7.0/gems/rspec-mocks-3.10.2/lib/rspec/mocks/message_expectation.rb:572:in `invoke_incrementing_actual_calls_by'
	from /home/runner/work/masterclass/masterclass/backend/vendor/bundle/ruby/2.7.0/gems/rspec-mocks-3.10.2/lib/rspec/mocks/message_expectation.rb:427:in `invoke'
	from /home/runner/work/masterclass/masterclass/backend/vendor/bundle/ruby/2.7.0/gems/rspec-mocks-3.10.2/lib/rspec/mocks/proxy.rb:214:in `message_received'
	from /home/runner/work/masterclass/masterclass/backend/vendor/bundle/ruby/2.7.0/gems/rspec-mocks-3.10.2/lib/rspec/mocks/proxy.rb:361:in `message_received'
	from /home/runner/work/masterclass/masterclass/backend/vendor/bundle/ruby/2.7.0/gems/rspec-mocks-3.10.2/lib/rspec/mocks/method_double.rb:78:in `proxy_method_invoked'
	from /home/runner/work/masterclass/masterclass/backend/vendor/bundle/ruby/2.7.0/gems/rspec-mocks-3.10.2/lib/rspec/mocks/method_double.rb:64:in `block (2 levels) in define_proxy_method'
	from /home/runner/work/masterclass/masterclass/backend/vendor/bundle/ruby/2.7.0/bundler/gems/test-collector-ruby-8074177c0faf/lib/buildkite/test_collector/session.rb:123:in `handle'
	from /home/runner/work/masterclass/masterclass/backend/vendor/bundle/ruby/2.7.0/bundler/gems/test-collector-ruby-8074177c0faf/lib/buildkite/test_collector/socket_connection.rb:72:in `block in initialize'
....................................................................................................*...................
```

Ahh, it looks like the "JSON Data" line is a `debug` log message, rather than `error`... https://github.com/buildkite/test-collector-ruby/blob/rdh-json-parse-error/lib/buildkite/test_collector/socket_connection.rb#L108-L109 

I guess we'll need to explicitly specify the log level to `Buildkite::TestCollector::Logger`?

**Comment by gchan** [comment]
Thanks for reporting this issue @rdh , you're right about the log level. I've changed it to `error` and updated the branch. 

Yes, you could also update the log level to `debug` as a workaround.

**Comment by rdh** [comment]
OK.  We caught one of these last night around 12:21am PT...

```
2022-08-04T07:19:49.564660420Z pid=7471 tid=5151520 ERROR: JSON::ParserError: JSON::ParserError
2022-08-04T07:19:49.568178053Z pid=7471 tid=5151520 ERROR: JSON data: {"type":"ping","message":1659597589}

#<Thread:0x000055e52e6fb4b8 /home/runner/work/masterclass/masterclass/backend/vendor/bundle/ruby/2.7.0/bundler/gems/test-collector-ruby-df6003b1f046/lib/buildkite/test_collector/socket_connection.rb:64 run> terminated with exception (report_on_exception is true):
```

I've done a grep for `ping` through our test code.  Found a lot of "shipping", "grouping", "mapping", even some legitimate `ping` methods.  But nothing mocking "ping" into a response.  I did a similar grep for `message` and did not find anything that pattern matched.

I'm doubly confused, as the JSON data appears to be well formed JSON that should not cause a `JSON::ParserError`.  Even copied the string into IRB to confirm that , to make sure my pre-coffee eyes weren't fooling me.

Taking a quick peek at this gem's source, I see we're logging `data`... 
https://github.com/buildkite/test-collector-ruby/blob/rdh-json-parse-error/lib/buildkite/test_collector/socket_connection.rb#L109

... while the failing code is actually receiving `data.data`...
https://github.com/buildkite/test-collector-ruby/blob/rdh-json-parse-error/lib/buildkite/test_collector/socket_connection.rb#L72

It also appears that `ping` is an expected message...
https://github.com/buildkite/test-collector-ruby/blob/rdh-json-parse-error/lib/buildkite/test_collector/session.rb#L125-L128

Without digging deeper into the code, my current guess is that the ping message is coming through as `data`, rather than `data.data`, and crashing.  But maybe `data` and `data.data` are identical, and something else is going on?  🤷‍♂️

**Comment by gchan** [comment]
Thanks for the write up @rdh 

We use ActionCable to support the streaming of data.

'ping' is a regular [heartbeat every 3 seconds](https://github.com/rails/rails/blob/8a52f9ba81ef31ac109e07678123182cdd14f52f/actioncable/lib/action_cable/server/connections.rb#L22-L24) to ensure the connection isn't stale.

From what I can tell at the moment, `handle` is only called in [one place](https://github.com/buildkite/test-collector-ruby/blob/rdh-json-parse-error/lib/buildkite/test_collector/socket_connection.rb#L72)  and always assumes `data` has a value for `data`. 

The `handle` [method](https://github.com/buildkite/test-collector-ruby/blob/df6003b1f0466c2749805843b8fd4ed63240a56b/lib/buildkite/test_collector/session.rb#L122) apparently handles the `ping` type but also `welcome`, `confirm_subscription`, and `reject_subscription`. 

Do you happen to have debug logging set up? 
Have you seen the following messages in your debug log?

`received rejected_subscription`
`received welcome`
`received confirm_subscription`

**Comment by gchan** [comment]
> {"type":"ping","message":1659620362}

This appears to be parse-able. Perhaps there is some hidden encoding we can't see?

Could you please check whether `JSON.parse` is being mocked? Perhaps search for `JSON::ParserError` in your codebase?

**Comment by rdh** [comment]
Aha!  I did find...
```
before { allow(JSON).to receive(:parse).and_raise(JSON::ParserError) }
```
... but this appears to be a completely legitimate usage to exercise our internal error handling.  I think you may need to mutex your thread so it's only running between tests, not while mocks are in effect.

**Comment by swebb** [comment]
I think we might be able to solve this issue in the same way as https://github.com/buildkite/test-collector-ruby/pull/132. We need to store a reference to JSON.parse and use that internally.

**Comment by gchan** [comment]
I've gone with the suggested approach of using a private method reference. The caveat is that there could be other methods that might need the same treatment.

I'm currently not sure if we can mutex the thread which handles the socket connection such that it only runs between tests. The purpose of the websocket implementation is so that we can stream test results to Buildkite.

**Comment by gchan** [comment]
@rdh thanks so much for your patience.

We believe we have resolved the issue in #149. You can two options to test the fix:

* Continue using the custom branch https://github.com/buildkite/test-collector-ruby/tree/rdh-json-parse-error . I've updated it to include the fix from the main branch
* Update your Gemfile to use the main branch

We will release a new version of this Ruby gem shortly. Thank you again!
